packages/@uppy/core/src/index.js

```diff
@@ -577,49 +577,143 @@ class Uppy {
         })
       }, 4)
     }
+  }
 
-    return fileID
+  /**
+   * Add a new file to `state.files`. This will run `onBeforeFileAdded`,
+   * try to guess file type in a clever way, check file against restrictions,
+   * and start an upload if `autoProceed === true`.
+   *
+   * @param {object} file object to add
+   * @returns {string} id for the added file
+   */
+  addFile (file) {
+    this._assertNewUploadAllowed(file)
+
+    const { files } = this.getState()
+    const newFile = this._checkAndCreateFileStateObject(files, file)
+
+    this.setState({
+      files: {
+        ...files,
+        [newFile.id]: newFile
+      }
+    })
+
+    this.emit('file-added', newFile)
+    this.log(`Added file: ${newFile.name}, ${newFile.id}, mime type: ${newFile.type}`)
+
+    this._startIfAutoProceed()
+
+    return newFile.id
   }
 
-  removeFile (fileID) {
+  /**
+   * Add multiple files to `state.files`. See the `addFile()` documentation.
+   *
+   * This cuts some corners for performance, so should typically only be used in cases where there may be a lot of files.
+   *
+   * If an error occurs while adding a file, it is logged and the user is notified. This is good for UI plugins, but not for programmatic use. Programmatic users should usually still use `addFile()` on individual files.
+   */
+  addFiles (fileDescriptors) {
+    this._assertNewUploadAllowed()
+
+    // create a copy of the files object only once
+    const files = { ...this.getState().files }
+    const newFiles = []
+    const errors = []
+    for (let i = 0; i < fileDescriptors.length; i++) {
+      try {
+        const newFile = this._checkAndCreateFileStateObject(files, fileDescriptors[i])
+        newFiles.push(newFile)
+        files[newFile.id] = newFile
+      } catch (err) {
+        if (!err.isRestriction) {
+          errors.push(err)
+        }
+      }
+    }
+
+    this.setState({ files })
+
+    newFiles.forEach((newFile) => {
+      this.emit('file-added', newFile)
+    })
+    this.log(`Added batch of ${newFiles.length} files`)
+
+    this._startIfAutoProceed()
+
+    if (errors.length > 0) {
+      const err = new Error('Multiple errors occured while adding files')
+      err.errors = errors
```

Comment:
Hm, if one error occurs after we try to add one file, - we will get a generic `Error: Multiple errors occured while adding files`, with no indication of what happened.
That could work if we were catching and logging `error.errors` everywhere after we use `.addFiles`, but we don't yet do that.

Should we log each error ourselves in `addFiles()` instead? That would also make:
```
    try {
      this.uppy.addFiles(descriptors)
    } catch (err) {
      this.uppy.log(err)
    }
```
tradition unnecessary?

PS. Looks like it's occurrRred (https://en.wiktionary.org/wiki/occured).